File: patterns/gate-keeper/rules-python/gate_keeper.py
import queue
import threading
from datetime import datetime, timedelta

from openhab import rule, Registry, logger
from openhab.actions import Exec
from openhab.triggers import ItemCommandTrigger

commands = queue.Queue()
gate_timer = None
last_command = datetime.now().astimezone() - timedelta(seconds=1)


def process_queue():
    global gate_timer, last_command
    if not commands.empty():
        cmd = commands.get()
        results = Exec.executeCommandLine(timedelta(seconds=5), cmd)
        logger.info("433: " + str(results))
        last_command = datetime.now().astimezone()

    delta_millis = (datetime.now().astimezone() - last_command).total_seconds() * 1000
    delay = (100 - delta_millis) / 1000 if delta_millis < 100 else 0
    gate_timer = threading.Timer(delay, process_queue)
    gate_timer.start()


@rule(triggers=[ItemCommandTrigger("WirelessController")])
class Controller433MHz:
    def execute(self, module, input):
        global gate_timer
        event = input.get("event")
        if not event:
            return

        commands.put(str(event.getItemCommand()))
        self.logger.info("Befehl in Warteschlange eingereiht: " + str(event.getItemCommand()))

        if gate_timer is None:
            process_queue()


@rule(triggers=[ItemCommandTrigger("Outlet_A")])
class OutletA:
    def execute(self, module, input):
        event = input.get("event")
        if not event:
            return

        if str(event.getItemCommand()) == "ON":
            Registry.getItem("WirelessController").sendCommand("433-send xxxxx 1 1")
        else:
            Registry.getItem("WirelessController").sendCommand("433-send xxxxx 1 0")

File: patterns/gate-keeper/rules-js/gate-keeper.js
const commands = [];
let gateTimer = null;
let lastCommand = time.ZonedDateTime.now().minusSeconds(1);

function processQueue() {
  if (commands.length > 0) {
    const cmd = commands.shift();
    const results = actions.Exec.executeCommandLine(time.Duration.ofMillis(5000), cmd.split(" "));
    console.debug("433: " + results);
    lastCommand = time.ZonedDateTime.now();
  }

  const deltaMillis = time.Duration.between(lastCommand, time.ZonedDateTime.now()).toMillis();
  gateTimer.reschedule(time.ZonedDateTime.now().plusMillis(deltaMillis < 100 ? 100 - deltaMillis : 0));
}

rules.JSRule({
  name: "433MHz Controller",
  triggers: [triggers.ItemCommandTrigger("WirelessController")],
  execute: (event) => {
    commands.push(String(event.receivedCommand));

    if (gateTimer === null) {
      gateTimer = actions.ScriptExecution.createTimer(time.ZonedDateTime.now(), processQueue);
    }
  }
});

rules.JSRule({
  name: "Outlet A",
  triggers: [triggers.ItemCommandTrigger("Outlet_A")],
  execute: (event) => {
    if (event.receivedCommand === "ON") {
      items.getItem("WirelessController").sendCommand("433-send xxxxx 1 1");
    } else {
      items.getItem("WirelessController").sendCommand("433-send xxxxx 1 0");
    }
  }
});
